File: auto_outsrc/parser/SW05/SW05_LATESTLATER.cpp
#include "Charm.h"
#include <iostream>
#include <sstream>
#include <string>
#include <list>
using namespace std;

PairingGroup group(AES_SECURITY);

SecretUtil util;

void setup(int n, CharmList & pk, ZR & mk)
{
    G1 g = group.init(G1_t);
    ZR y = group.init(ZR_t);
    G1 g1 = group.init(G1_t);
    G2 g2 = group.init(G2_t);
    CharmListG2 t;
    g = group.random(G1_t);
    y = group.random(ZR_t);
    g1 = group.exp(g, y);
    g2 = group.random(G2_t);
    for (int i = 0; i < n+1; i++)
    {
        t.insert(i, group.random(G2_t));
    }
    pk.insert(0, g);
    pk.insert(1, g1);
    pk.insert(2, g2);
    pk.insert(3, t);
    mk = y;
    return;
}

void evalT(CharmList & pk, int n, ZR & x, G2 & T)
{
    G1 g;
    G1 g1;
    G2 g2;
    CharmListG2 t;
    CharmListInt N;
    CharmDictZR coeffs;
    G2 prodResult = group.init(G2_t);
    int lenN = 0;
    int loopVarEvalT = 0;
    int loopVarM1 = 0;
    
    g = pk[0].getG1();
    g1 = pk[1].getG1();
    g2 = pk[2].getG2();
    t = pk[3].getListG2();
    for (int i = 0; i < n+1; i++)
    {
        N.insert(i, group.add(i, 1));
    }
    coeffs = util.recoverCoefficientsDict(group, N);
    //;
    lenN = N.length();
    for (int i = 0; i < lenN; i++)
    {
        loopVarEvalT = N[i];
        loopVarM1 = group.sub(loopVarEvalT, 1);
        prodResult = group.mul(prodResult, group.exp(t[loopVarM1], coeffs[loopVarEvalT]));
    }
    T = group.mul(group.exp(g2, group.mul(x, n)), prodResult);
    return;
}

void extract(ZR & mk, CharmListStr & w, CharmList & pk, int dParam, int n, ZR & bf0, CharmList & skBlinded)
{
    string wBlinded;
    G1 g;
    G1 g1;
    G2 g2;
    CharmListG2 t;
    int lenw = 0;
    string loopVar1;
    CharmListZR wHash;
    ZR r = group.init(ZR_t);
    CharmListZR q;
    CharmListZR shares;
    int wHashLen = 0;
    ZR loopVar2 = group.init(ZR_t);
    string loopVar2Str;
    G2 evalTVar = group.init(G2_t);
    CharmListG2 D;
    CharmListG1 d;
    CharmListZR blindingFactordBlinded;
    CharmListG1 dBlinded;
    CharmListZR blindingFactorDBlinded;
    CharmListG2 DBlinded;
    bf0 = group.random(ZR_t);
    wBlinded = w;
    
    g = pk[0].getG1();
    g1 = pk[1].getG1();
    g2 = pk[2].getG2();
    t = pk[3].getListG2();
    lenw = wBlinded.length();
    for (int i = 0; i < lenw; i++)
    {
        loopVar1 = wBlinded[i];
        wHash.insert(i, group.hashListToZR(loopVar1));
    }
    r = group.random(ZR_t);
    q.insert(0, mk);
    for (int i = 1; i < dParam; i++)
    {
        q.insert(i, group.random(ZR_t));
    }
    shares = genSharesForX(mk, q, wHash);
    wHashLen = wHash.length();
    for (int i = 0; i < wHashLen; i++)
    {
        loopVar2 = wHash[i];
        loopVar2Str = wBlinded[i];
        evalTVar = evalT(pk, n, loopVar2);
        D.insert(loopVar2Str, group.mul(group.exp(g2, shares[i]), group.exp(evalTVar, r)));
        d.insert(loopVar2Str, group.exp(g, r));
    }
    CharmList d_keys = d.keys();
    int d_len = d.length();
    for (int y_var = 0; y_var < d_len; y_var++)
    {
        y = d_keys[y_var];
        blindingFactordBlinded.insert(y, bf0);
        dBlinded.insert(y, group.exp(d[y], group.div(1, blindingFactordBlinded[y].getZR())));
    }
    CharmList D_keys = D.keys();
    int D_len = D.length();
    for (int y_var = 0; y_var < D_len; y_var++)
    {
        y = D_keys[y_var];
        blindingFactorDBlinded.insert(y, bf0);
        DBlinded.insert(y, group.exp(D[y], group.div(1, blindingFactorDBlinded[y].getZR())));
    }
    skBlinded.insert(0, wBlinded);
    skBlinded.insert(1, DBlinded);
    skBlinded.insert(2, dBlinded);
    return;
}

void encrypt(CharmList & pk, CharmListStr & wPrime, GT & M, int n, CharmList & CT)
{
    G1 g;
    G1 g1;
    G2 g2;
    CharmListG2 t;
    ZR s = group.init(ZR_t);
    GT Eprime = group.init(GT_t);
    G1 Eprimeprime = group.init(G1_t);
    int wPrimeLen = 0;
    ZR loopVar = group.init(ZR_t);
    string loopVarStr;
    G2 evalTVar = group.init(G2_t);
    CharmListG2 E;
    
    g = pk[0].getG1();
    g1 = pk[1].getG1();
    g2 = pk[2].getG2();
    t = pk[3].getListG2();
    s = group.random(ZR_t);
    Eprime = group.mul(M, group.exp(group.pair(g1, g2), s));
    Eprimeprime = group.exp(g, s);
    wPrimeLen = wPrime.length();
    for (int i = 0; i < wPrimeLen; i++)
    {
        loopVar = group.hashListToZR(wPrime[i]);
        loopVarStr = wPrime[i];
        evalTVar = evalT(pk, n, loopVar);
        E.insert(loopVarStr, group.exp(evalTVar, s));
    }
    CT.insert(0, wPrime);
    CT.insert(1, Eprime);
    CT.insert(2, Eprimeprime);
    CT.insert(3, E);
    return;
}

void transform(CharmList & pk, CharmList & skBlinded, CharmList & CT, int dParam, CharmList & transformOutputList, CharmListStr & SKeys, int SLen)
{

File: auto_outsrc/parser/SW05/SW05_LATESTLATER.py
from SW05_LATESTLATER_USER import *

from charm.toolbox.pairinggroup import PairingGroup,ZR,G1,G2,GT,pair
from charm.core.engine.util import *
from charm.core.math.integer import randomBits

group = None

secparam = 80


def setup(n):
    t = {}

    g = group.random(G1)
    y = group.random(ZR)
    g1 = (g ** y)
    g2 = group.random(G2)
    for i in range(0, n+1):
        t[i] = group.random(G2)
    pk = [g, g1, g2, t]
    mk = y
    output = (pk, mk)
    return output

def evalT(pk, n, x):
    N = {}

    g, g1, g2, t = pk
    for i in range(0, n+1):
        N[i] = (i + 1)
    coeffs = recoverCoefficientsDict(N)
    prodResult = group.init(G2)
    lenN = len(N)
    for i in range(0, lenN):
        loopVarEvalT = N[i]
        loopVarM1 = (loopVarEvalT - 1)
        prodResult = (prodResult * (t[loopVarM1] ** coeffs[loopVarEvalT]))
    T = ((g2 ** (x * n)) * prodResult)
    output = T
    return output

def extract(mk, w, pk, dParam, n):
    DBlinded = {}
    blindingFactorDBlinded = {}
    q = {}
    D = {}
    d = {}
    dBlinded = {}
    wHash = {}
    blindingFactordBlinded = {}

    bf0 = group.random(ZR)
    wBlinded = w
    g, g1, g2, t = pk
    lenw = len(wBlinded)
    for i in range(0, lenw):
        loopVar1 = wBlinded[i]
        wHash[i] = group.hash(loopVar1, ZR)
    r = group.random(ZR)
    q[0] = mk
    for i in range(1, dParam):
        q[i] = group.random(ZR)
    shares = genSharesForX(mk, q, wHash)
    wHashLen = len(wHash)
    for i in range(0, wHashLen):
        loopVar2 = wHash[i]
        loopVar2Str = wBlinded[i]
        evalTVar = evalT(pk, n, loopVar2)
        D[loopVar2Str] = ((g2 ** shares[i]) * (evalTVar ** r))
        d[loopVar2Str] = (g ** r)
    for y in d:
        blindingFactordBlinded[y] = bf0
        dBlinded[y] = (d[y] ** (1 / blindingFactordBlinded[y]))
    for y in D:
        blindingFactorDBlinded[y] = bf0
        DBlinded[y] = (D[y] ** (1 / blindingFactorDBlinded[y]))
    skBlinded = [wBlinded, DBlinded, dBlinded]
    output = (bf0, skBlinded)
    return output

def encrypt(pk, wPrime, M, n):
    E = {}

    g, g1, g2, t = pk
    s = group.random(ZR)
    Eprime = (M * (pair(g1, g2) ** s))
    Eprimeprime = (g ** s)
    wPrimeLen = len(wPrime)
    for i in range(0, wPrimeLen):
        loopVar = group.hash(wPrime[i], ZR)
        loopVarStr = wPrime[i]
        evalTVar = evalT(pk, n, loopVar)
        E[loopVarStr] = (evalTVar ** s)
    CT = [wPrime, Eprime, Eprimeprime, E]
    output = CT
    return output

def transform(pk, skBlinded, CT, dParam):
    transformOutputList = {}

    wPrime, Eprime, Eprimeprime, E = CT
    wBlinded, DBlinded, dBlinded = skBlinded
    transformOutputList[1] = Eprime
    transformOutputList[0] = intersectionSubset(wBlinded, wPrime, dParam)
    S = transformOutputList[0]
    coeffs = recoverCoefficientsDict(S)
    SKeys = strkeys(S)
    SLen = len(S)
    for i in range(0, SLen):
        pass
        loopVar3 = SKeys[i]
        transformOutputList[1000+5*i] = (pair((dBlinded[loopVar3] ** coeffs[loopVar3]), E[loopVar3]) * pair((Eprimeprime ** -coeffs[loopVar3]), DBlinded[loopVar3]))
        loopProd = transformOutputList[1000+5*i]
    output = (transformOutputList, SKeys, SLen)
    return output

def decout(pk, dParam, transformOutputList, bf0, SKeys, SLen):
    Eprime = transformOutputList[1]
    S = transformOutputList[0]
    prod = group.init(GT)
    for i in range(0, SLen):
        pass
        loopVar3 = SKeys[i]
        loopProd = (transformOutputList[1000+5*i] ** bf0)
        prod = (prod * loopProd)
    M = (Eprime * prod)
    output = M
    return output

def SmallExp(bits=80):
    return group.init(ZR, randomBits(bits))

def main():
    global group
    group = PairingGroup("SS512")

    max_attributes = 6
    required_overlap = 4
    #(master_public_key, master_key) = setup(max_attributes, required_overlap)
    (master_public_key, master_key) = setup(max_attributes)
    private_identity = ['insurance', 'id=2345', 'oncology', 'doctor', 'nurse', 'JHU'] #private identity
    public_identity = ['insurance', 'id=2345', 'doctor', 'oncology', 'JHU', 'billing', 'misc'] #public identity for encrypt
    (blindingFactor0Blinded, skBlinded) = extract(master_key, private_identity, master_public_key, required_overlap, max_attributes)
    msg = group.random(GT)
    cipher_text = encrypt(master_public_key, public_identity, msg, max_attributes)
    #decrypted_msg = decrypt(master_public_key, secret_key, cipher_text, required_overlap)
    (transformOutputList, SKeys, SLen) = transform(master_public_key, skBlinded, cipher_text, required_overlap)
    decrypted_msg = decout(master_public_key, required_overlap, transformOutputList, blindingFactor0Blinded, SKeys, SLen)

    print("msg:  ", msg)
    print("decrypted_msg:  ", decrypted_msg)
    assert msg == decrypted_msg, "failed decryption!"
    print("Successful Decryption!")


if __name__ == '__main__':
    main()

